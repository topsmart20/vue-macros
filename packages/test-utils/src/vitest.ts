import path from 'node:path'
import glob, { type Options as GlobOptions } from 'fast-glob'
import { describe, expect, test } from 'vitest'
import { normalizePath } from '@rollup/pluginutils'

interface Options {
  params?: [name: string, values?: any[]][]
  promise?: boolean
}

const SKIP_VUE2 = !!process.env.SKIP_VUE2

export async function testFixtures(
  globs: string | string[],
  exec: (args: Record<string, any>, id: string) => any,
  options?: GlobOptions & Options,
): Promise<void>
export async function testFixtures(
  files: Record<string, string>,
  exec: (args: Record<string, any>, id: string, code: string) => any,
  options?: Options,
): Promise<void>
export async function testFixtures(
  globsOrFiles: string | string[] | Record<string, string>,
  cb: (args: Record<string, any>, id: string, code: string) => any,
  { params, promise, ...globOptions }: GlobOptions & Options = {},
) {
  let files: Record<string, string | undefined>

  if (typeof globsOrFiles === 'string' || Array.isArray(globsOrFiles)) {
    files = Object.fromEntries(
      (await glob(globsOrFiles, globOptions)).map((file) => [file, undefined]),
    )
  } else {
    files = globsOrFiles
  }

  for (const [id, code] of Object.entries(files)) {
    makeTests(id, code, [[normalizePath(id)], ...(params || [])])()
  }

  function getName(name: string, value: any) {
    return value !== undefined ? `${name} = ${String(value)}` : name
  }

  function makeTests(
    id: string,
    code: string | undefined,
    params: NonNullable<Options['params']>,
    args: Record<string, any> = {},
  ) {
    const [currParams, ...restParams] = params
    const [name, values = [undefined]] = currParams
    if (restParams.length > 0) {
      return () => {
        for (const value of values) {
          const currArgs = { ...args, [name]: value }

          describe(
            getName(name, value),
            makeTests(id, code, restParams, currArgs),
          )
        }
      }
    } else {
      return () => {
        for (const value of values) {
          const testName = getName(name, value)
          const isSkip = testName.includes('vue2') && SKIP_VUE2

          test.skipIf(isSkip)(testName, async () => {
            const currArgs = { ...args, [name]: value }
            const execute = () =>
              cb(currArgs, path.resolve(globOptions.cwd || '/', id), code!)
            if (id.includes('error')) {
              if (promise) {
                await expect(execute()).rejects.toThrowErrorMatchingSnapshot()
              } else {
                expect(execute).toThrowErrorMatchingSnapshot()
              }
            } else if (promise) {
              await expect(
                (execute() as Promise<any>).catch((error) => {
                  console.warn(error)
                  return Promise.reject(error)
                }),
              ).resolves.toMatchSnapshot()
            } else {
              expect(execute()).toMatchSnapshot()
            }
          })
        }
      }
    }
  }
}
