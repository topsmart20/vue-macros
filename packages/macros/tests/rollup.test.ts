import { resolve } from 'node:path'
import { describe, expect, it } from 'vitest'
import glob from 'fast-glob'
import VueMacros from '../src/rollup'
import { ToString, getCode } from './_utils'

describe('Rollup', () => {
  describe('fixtures', async () => {
    const root = resolve(__dirname, '..')
    const files = await glob('tests/fixtures/*.{vue,js,ts}', {
      cwd: root,
      onlyFiles: true,
    })

    for (const file of files) {
      it(file.replace(/\\/g, '/'), async () => {
        const filepath = resolve(root, file)
        const version = filepath.includes('vue2') ? 2 : 3

        const unpluginCode = await getCode(filepath, [
          VueMacros({ version }),
          ToString,
        ])
        expect(unpluginCode).toMatchSnapshot()
      })
    }
  })
})
