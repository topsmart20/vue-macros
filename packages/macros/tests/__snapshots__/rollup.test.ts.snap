// Vitest Snapshot v1

exports[`Rollup > fixtures > tests/fixtures/mixed.vue 1`] = `
"var mixed = \`<script lang=\\"ts\\">
import { defineComponent as DO_defineComponent } from 'vue';
export default /*#__PURE__*/ DO_defineComponent({
  name: 'Foo',
});
</script>
<script setup lang=\\"ts\\">import { emitHelper as __emitHelper } from 'unplugin-vue-macros/helper'


const {modelValue,  title } = defineProps<{modelValue: string

  title: string
}>()
const emit = defineEmits<{(evt: 'update:modelValue', value: string): void

  (evt: 'change'): void
}>()


const handleClick = () => {
  emit('change')
  __emitHelper(emit, 'update:modelValue', 'hello, ' + title)
}
</script>
\`;

export { mixed as default };
"
`;

exports[`Rollup > fixtures > tests/fixtures/mixed-vue2-model.vue 1`] = `
"var mixedVue2Model = \`<script lang=\\"ts\\">
import { defineComponent as DO_defineComponent } from 'vue';
export default /*#__PURE__*/ DO_defineComponent({
  name: 'Foo',
  model: {
    prop: 'value',
    event: 'changeInput'
  },
});
</script>
<script setup lang=\\"ts\\">import { emitHelper as __emitHelper } from 'unplugin-vue-macros/helper'


const {modelValue, value,  title } = defineProps<{modelValue: string
value: string

  title: string
}>()
const emit = defineEmits<{(evt: 'update:modelValue', value: string): void
(evt: 'changeInput', value: string): void

  (evt: 'change'): void
}>()


const handleClick = () => {
  emit('change')
  __emitHelper(emit, 'update:modelValue', 'hello, ' + title)
  __emitHelper(emit, 'changeInput', 'Word,' + title)
}
</script>
\`;

export { mixedVue2Model as default };
"
`;
